cpp/include/raft/core/resource/device_memory_resource.hpp

```diff
@@ -40,36 +65,109 @@ class device_memory_resource : public resource {
  */
 class workspace_resource_factory : public resource_factory {
  public:
-  workspace_resource_factory(rmm::mr::device_memory_resource* mr_ = nullptr) : mr(mr_) {}
-  resource_type get_resource_type() override { return resource_type::WORKSPACE_RESOURCE; }
-  resource* make_resource() override { return new device_memory_resource(mr); }
+  explicit workspace_resource_factory(
+    std::shared_ptr<rmm::mr::device_memory_resource> mr = {nullptr},
+    std::optional<std::size_t> allocation_limit         = std::nullopt,
+    std::optional<std::size_t> alignment                = std::nullopt)
+    : allocation_limit_(allocation_limit.value_or(default_allocation_limit())),
+      alignment_(alignment),
+      mr_(mr ? mr : default_memory_resource(allocation_limit_))
+  {
+  }
+
+  auto get_resource_type() -> resource_type override { return resource_type::WORKSPACE_RESOURCE; }
+  auto make_resource() -> resource* override
+  {
+    return new limited_memory_resource(mr_, allocation_limit_, alignment_);
+  }
 
  private:
-  rmm::mr::device_memory_resource* mr;
+  std::size_t allocation_limit_;
+  std::optional<std::size_t> alignment_;
+  std::shared_ptr<rmm::mr::device_memory_resource> mr_;
+
+  // Create a pool memory resource by default
+  static inline auto default_memory_resource(std::size_t limit)
+    -> std::shared_ptr<rmm::mr::device_memory_resource>
+  {
+    constexpr std::size_t kOneGb = 1024lu * 1024lu * 1024lu;
+    auto min_size                = std::min<std::size_t>(kOneGb, limit / 2);
+    auto max_size                = limit * 3lu / 2lu;
+    auto upstream                = rmm::mr::get_current_device_resource();
+    return std::make_shared<rmm::mr::pool_memory_resource<rmm::mr::device_memory_resource>>(
+      upstream, min_size, max_size);
+  }
+
+  // Allow a fraction of available memory by default.
+  static inline auto default_allocation_limit() -> std::size_t
+  {
+    std::size_t free_size{};
+    std::size_t total_size{};
+    RAFT_CUDA_TRY(cudaMemGetInfo(&free_size, &total_size));
+    return free_size / 2;
```

Comment:
Note that using `free_size` here is probably not what we want: if an upstream pooling allocator has already grabbed the space, that would be misleading. 
- One could use `total_size`, to set a practically "unlimited" workspace. This is consistent with the existing not limited workspace allocator.  But then we could not rely on the workspace size to calculate batch size without risk of running OOM. 
- That brings me back to the point of allowing a smaller limited workspace size, that is consistent with the algorithm hardcoded heuristics.
- If I see it correctly, the workspace resource is now used for all mdarrays https://github.com/rapidsai/raft/blob/e963f5a64371c25f1e2e09ab1e3ca15658293e2d/cpp/include/raft/core/device_container_policy.hpp#L167  Does this mean that an ivf_pq::index is allocating its device_matrix from the workspace? @cjnolet

Reply:
> the workspace resource is now used for all mdarrays

Oh. Thanks for pointing this out! I think, the meanings we put into the workspace resource start to diverge again. 

With the change to container policy you quoted, the workspace resource essentially becomes a raft substitute to `rmm::mr::current_device_resource`, because the policy is used for allocating all mdarrays, including the ones passed to the user inside structs like `ivf_*::index`. If we're to hold on to this approach, this PR certainly does not make any sense at all.

In this PR we assume the workspace resource is used only for the temporary allocations within the algorithms. If we are to proceed with the PR, we must revert the container policy change. I also believe the default pool allocator is justified in this case, because (1) it's good for performance, (2) it does not affect the allocations outside its explicit use and thus does not conflict with other (e.g. third-party) code, (3) we can enforce a conservative limit to avoid taking up all gpu resources.

Reply:
Sorry Artem, I'm still catching up on this discussion a bit. From what I've read so far, I think your design is headed in the right direction but I think there's still a few things to polish off (like the defaults, for example).